# Modelo de predicción de inscripciones estudiantiles mediante técnicas de Machine Learning

**Maestría en Inteligencia Artificial Aplicada**

**Proyecto Integrador Sep-Nov 2025**

**Equipo 14**

**Avance 4.    Modelos Alternativos**

**Integrantes:**

- Alejandro Roa Solis – A01129942
- Annette Cristina Narvaez Andrade – A00571041
- Karla Alejandra Fonseca Márquez – A01795313


**Patrocinador Tec de Monterrey:**

Dr. Juan Arturo Nolazco Flores, Director del Hub de Ciencias y Datos de la Escuela de Ingeniería del Tec de Monterrey




# Avance 4 — Modelos Alternativos (por subconjunto)

En esta cuarta etapa del *Proyecto Integrador* se desarrolla la **fase de modelado**, correspondiente a la metodología **CRISP-ML (Cross-Industry Standard Process for Machine Learning)**.  

Su propósito es **construir, comparar y ajustar múltiples modelos predictivos individuales**, con el fin de identificar cuál ofrece el mejor desempeño para el problema de predicción de inscripción estudiantil.

A diferencia del avance anterior —donde se estableció un *modelo baseline*—, en esta fase se implementan **diversos algoritmos de clasificación supervisada**, cada uno con configuraciones iniciales propias.  
Esta diversidad permite explorar la capacidad de los datos para adaptarse a distintas estructuras de aprendizaje y analizar la sensibilidad del rendimiento frente a variaciones de hiperparámetros.

---

Objetivos del avance

1. **Entrenar al menos seis modelos individuales**, utilizando algoritmos representativos de distintas familias (lineales, basados en árboles, de márgenes, entre otros).  
2. **Evaluar y comparar el desempeño** de los modelos mediante métricas de clasificación relevantes (*accuracy, F1-score, matriz de confusión, validación cruzada*).  
3. **Seleccionar los dos modelos con mejores resultados** y aplicar **técnicas de ajuste de hiperparámetros** (*hyperparameter tuning*) para optimizar su rendimiento.  
4. **Elegir el modelo individual final**, considerando desempeño cuantitativo, interpretabilidad, robustez y factibilidad de implementación.

---

Con este avance se consolida el proceso de **experimentación y refinamiento de los modelos**, garantizando la construcción de una solución predictiva sólida, evaluada bajo criterios reproducibles y alineados con los objetivos de negocio.


In [3]:
# Librerías y utilidades

import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from typing import Dict, Tuple, Callable

import os
import joblib
from joblib import dump

# Modelado y validación
from scipy.stats import randint, uniform
from sklearn.model_selection import (
    train_test_split, StratifiedKFold,
    cross_val_score, RandomizedSearchCV,
    GridSearchCV
)

# Preprocesamiento y pipelines
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

# Modelos
from sklearn.linear_model import LogisticRegression, RidgeClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.naive_bayes import GaussianNB
from sklearn.neural_network import MLPClassifier

# Métricas
from sklearn.metrics import (
    accuracy_score, f1_score, roc_auc_score,
    confusion_matrix, classification_report
)
from sklearn.utils.class_weight import compute_class_weight


In [4]:
# Definición e inicialización de variables

RANDOM_STATE = 42
TEST_SIZE = 0.30
TARGET_COL = "INSCRITO"

os.makedirs("modelos", exist_ok=True)

In [5]:
# Definicion de fuentes de datos, una para cada subconjunto
SUBSETS: Dict[str, dict] = {
    "PrepaTec": {"csv": "FE_DS/prepa_tec_feature_engineered_full.csv"},
    "Profesional_PrepaTec": {"csv": "FE_DS/profesional_tec_feature_engineered_full.csv"},
    "Profesional_Externos": {"csv": "FE_DS/profesional_nnotec_feature_engineered_full.csv"},
}

In [6]:
# Función para cargar los datasets existentes
def load_subset(name: str, cfg: dict) -> Tuple[pd.DataFrame, pd.Series]:
    """Carga X, y para un subconjunto.

    Soporta:
      - cfg = {"csv": "path.csv"}  (usa todo el archivo)
      - cfg = {"csv": "path.csv", "filter": {"COL": "VAL", ...}} (filtra antes de separar)
    """
    path = cfg.get("csv")
    if not path or not os.path.exists(path):
        raise FileNotFoundError(f"[{name}] No se encontró el CSV: {path}")

    df = pd.read_csv(path)
    if "filter" in cfg and isinstance(cfg["filter"], dict):
        for col, val in cfg["filter"].items():
            if col not in df.columns:
                raise KeyError(f"[{name}] La columna de filtro '{col}' no existe en el CSV.")
            df = df[df[col] == val]

    if TARGET_COL not in df.columns:
        raise KeyError(f"[{name}] No se encontró la columna objetivo '{TARGET_COL}'.")

    y = df[TARGET_COL]
    X = df.drop(columns=[TARGET_COL])

    # Seguridad: si hay columnas completamente vacías, descartarlas
    empty_cols = [c for c in X.columns if X[c].isna().all()]
    if empty_cols:
        X = X.drop(columns=empty_cols)

    return X, y

In [7]:
from sklearn.exceptions import NotFittedError

def get_y_score(model, X_test):
    try:
        # Si el modelo tiene método predict_proba, usarlo
        if hasattr(model, "predict_proba"):
            probs = model.predict_proba(X_test)
            # Si devuelve matriz de 2 columnas, tomar probabilidad de clase positiva
            if probs.shape[1] == 2:
                return probs[:, 1]
            else:
                # Si es multiclase, puedes usar la clase más probable
                return probs.max(axis=1)
        
        # Si no, usar decision_function si está disponible
        elif hasattr(model, "decision_function"):
            return model.decision_function(X_test)
        
        else:
            raise AttributeError("El modelo no tiene ni predict_proba ni decision_function.")
    
    except NotFittedError:
        raise RuntimeError("El modelo no ha sido entrenado. Debe ajustarse antes de calcular y_score.")


In [8]:
def train_evaluate_cv(model, name):
    model.fit(X_tr, y_tr)

    # Evaluación en test
    y_pred = model.predict(X_te)

    y_score = get_y_score(model, X_te)
    
    acc = accuracy_score(y_te, y_pred)
    f1_bin = f1_score(y_te, y_pred)
    f1_w = f1_score(y_te, y_pred, average="weighted")
    roc = roc_auc_score(y_te, y_score)
       
    
    print(f"[{subset}] {name} — Test")
    print(f"Accuracy:        {acc:.3f}")
    print(f"F1 (binary):     {f1_bin:.3f}")
    print(f"F1 (weighted):   {f1_w:.3f}")
    print(f"ROC-AUC:         {roc:.3f}")
    print("\nMatriz de confusión:\n", confusion_matrix(y_te, y_pred))
    print("\nClassification report:\n", classification_report(y_te, y_pred, zero_division=0))
    
    # Validación cruzada (macro o weighted para desbalance)
    cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)
    cv_f1w = cross_val_score(model, X, y, scoring="f1_weighted", cv=cv, n_jobs=-1)
    print(f"\nCV F1-weighted 5-fold: {cv_f1w.mean():.3f} ± {cv_f1w.std():.3f}") 


## Selección de modelos

En esta etapa se implementaron y evaluaron **seis modelos individuales de clasificación supervisada**, adicionales al baseline, con el objetivo de identificar qué algoritmos se adaptan mejor a las características de cada subconjunto de datos.  

Cada modelo fue entrenado y probado utilizando una **división estratificada (70/30)**, manteniendo la proporción original de clases y controlando la aleatoriedad mediante una semilla fija (`random_state=42`).

Dado que los subconjuntos presentan un **alto desbalance de clases**, se priorizaron métricas más robustas que la exactitud, como el **F1-score ponderado (weighted)** y el **ROC-AUC**, complementadas con validación cruzada de 5 pliegues para evaluar la estabilidad del rendimiento.

Los modelos seleccionados representan diferentes familias de algoritmos, permitiendo contrastar sus fortalezas:

| **Modelo** | **Tipo / Familia** | **Características principales** |
|-------------|--------------------|---------------------------------|
| **Logistic Regression (Balanceada)** *Baseline* | Lineal / clasificación probabilística | Estima la probabilidad de que una observación pertenezca a una clase mediante una función sigmoide. Es uno de los modelos más interpretables, eficiente con grandes volúmenes de datos y útil cuando la relación entre variables y salida es aproximadamente lineal. |
| **Support Vector Machine (RBF)** | No lineal / separación de márgenes | Busca un hiperplano que maximiza la separación entre clases. Mediante el uso de funciones kernel (como RBF), puede capturar fronteras de decisión no lineales en espacios transformados. Es robusto ante sobreajuste, pero computacionalmente costoso en conjuntos grandes. |
| **K-Nearest Neighbors** | Basado en instancia / aprendizaje perezoso | Clasifica un nuevo dato en función de las clases de sus k vecinos más cercanos en el espacio de características. No requiere entrenamiento explícito, pero su rendimiento depende de la distancia elegida y la escala de los datos. |
| **Decision Tree** | Basado en reglas / aprendizaje jerárquico | Crea un árbol de decisiones dividiendo los datos en nodos según la ganancia de información o la reducción de impureza. Es interpretable, puede manejar relaciones no lineales y variables categóricas, aunque propenso al sobreajuste si no se regula la profundidad.|
| **Naive Bayes (Gaussian)** | Probabilístico / enfoque bayesiano | Aplica el teorema de Bayes asumiendo independencia condicional entre las variables predictoras. Es rápido, simple y funciona bien en textos o datos de alta dimensión, aunque suponer independencia puede limitar su desempeño en variables correlacionadas. |
| **MLPClassifier (Red Neuronal Multicapa)** | Conexionista / aprendizaje profundo supervisado | Se compone de capas de neuronas artificiales conectadas mediante funciones de activación no lineales. Aprende representaciones complejas y relaciones no lineales entre variables. Requiere ajuste cuidadoso de hiperparámetros y suficiente cantidad de datos. |
| **RidgeClassifier** | Lineal regularizado / Clasificación | Variante de la regresión lineal que introduce una penalización L2 sobre los coeficientes para reducir el sobreajuste. Es estable, eficiente y adecuado cuando existen variables correlacionadas o de alta dimensionalidad. |

Cada modelo fue evaluado con métricas de desempeño en el conjunto de prueba (**Accuracy**, **F1-score**, **ROC-AUC**) y mediante **validación cruzada**, generando una interpretación individual que analiza su comportamiento frente al desbalance y su potencial de generalización.


## PrepaTec

In [11]:
# 1) Cargar datos del subset
subset = "PrepaTec"
X, y = load_subset(subset, SUBSETS[subset])

# 2) Split estratificado
X_tr, X_te, y_tr, y_te = train_test_split(
    X, y, test_size=TEST_SIZE, random_state=RANDOM_STATE, stratify=y
)

In [12]:
# ===== Modelo 1: Logistic Regression (Balanced) =====

# Definir modelo Logistic Regression (balanceado)
logreg = LogisticRegression(
    multi_class="ovr",
    max_iter=2000,
    class_weight="balanced",     # compensa desbalance de clases
    random_state=RANDOM_STATE
)

train_evaluate_cv(logreg, "Logistic Regression")

[PrepaTec] Logistic Regression — Test
Accuracy:        0.581
F1 (binary):     0.686
F1 (weighted):   0.627
ROC-AUC:         0.663

Matriz de confusión:
 [[ 354  176]
 [1025 1309]]

Classification report:
               precision    recall  f1-score   support

           0       0.26      0.67      0.37       530
           1       0.88      0.56      0.69      2334

    accuracy                           0.58      2864
   macro avg       0.57      0.61      0.53      2864
weighted avg       0.77      0.58      0.63      2864


CV F1-weighted 5-fold: 0.628 ± 0.009


In [13]:
# ===== Modelo 2: Support Vector Classifier (SVC-RBF) =====

# Definir pipeline: SVC
svc_pipe = Pipeline([
    ("svc", SVC(
        kernel="rbf",
        C=1,
        gamma="scale",
        probability=True,       # necesario para ROC-AUC
        class_weight="balanced",
        random_state=RANDOM_STATE
    ))
])

train_evaluate_cv(svc_pipe, "SVC (RBF)")

[PrepaTec] SVC (RBF) — Test
Accuracy:        0.596
F1 (binary):     0.701
F1 (weighted):   0.641
ROC-AUC:         0.662

Matriz de confusión:
 [[ 355  175]
 [ 981 1353]]

Classification report:
               precision    recall  f1-score   support

           0       0.27      0.67      0.38       530
           1       0.89      0.58      0.70      2334

    accuracy                           0.60      2864
   macro avg       0.58      0.62      0.54      2864
weighted avg       0.77      0.60      0.64      2864


CV F1-weighted 5-fold: 0.632 ± 0.017


In [14]:
# ===== Modelo 3: KNN =====

# Definir pipeline: KNN (config inicial razonable)
knn_pipe = Pipeline([
    ("knn", KNeighborsClassifier(
        n_neighbors=11,     # k impar para evitar empates
        weights="distance", # vecinos más cercanos pesan más
        p=2,                # distancia Euclidiana (prueba p=1 también)
        n_jobs=-1
    ))
])

train_evaluate_cv(knn_pipe, "KNN")

[PrepaTec] KNN — Test
Accuracy:        0.811
F1 (binary):     0.894
F1 (weighted):   0.749
ROC-AUC:         0.625

Matriz de confusión:
 [[  34  496]
 [  45 2289]]

Classification report:
               precision    recall  f1-score   support

           0       0.43      0.06      0.11       530
           1       0.82      0.98      0.89      2334

    accuracy                           0.81      2864
   macro avg       0.63      0.52      0.50      2864
weighted avg       0.75      0.81      0.75      2864


CV F1-weighted 5-fold: 0.743 ± 0.003


In [15]:
# ===== Modelo 4: Decision Tree =====

# Árbol base con poda ligera
dt = DecisionTreeClassifier(
    criterion="gini",
    max_depth=None,          # deja crecer y que ccp_alpha pode
    min_samples_split=10,
    min_samples_leaf=5,
    class_weight="balanced",
    random_state=RANDOM_STATE
)

train_evaluate_cv(dt, "DecisionTree")

[PrepaTec] DecisionTree — Test
Accuracy:        0.629
F1 (binary):     0.751
F1 (weighted):   0.662
ROC-AUC:         0.544

Matriz de confusión:
 [[ 198  332]
 [ 731 1603]]

Classification report:
               precision    recall  f1-score   support

           0       0.21      0.37      0.27       530
           1       0.83      0.69      0.75      2334

    accuracy                           0.63      2864
   macro avg       0.52      0.53      0.51      2864
weighted avg       0.71      0.63      0.66      2864


CV F1-weighted 5-fold: 0.666 ± 0.007


In [16]:
# ===== Modelo 5: Naive Bayes (GaussianNB) =====

# Entrenar modelo GaussianNB
nb = GaussianNB()

train_evaluate_cv(nb, "NaiveBayes")

[PrepaTec] NaiveBayes — Test
Accuracy:        0.272
F1 (binary):     0.205
F1 (weighted):   0.228
ROC-AUC:         0.635

Matriz de confusión:
 [[ 509   21]
 [2065  269]]

Classification report:
               precision    recall  f1-score   support

           0       0.20      0.96      0.33       530
           1       0.93      0.12      0.21      2334

    accuracy                           0.27      2864
   macro avg       0.56      0.54      0.27      2864
weighted avg       0.79      0.27      0.23      2864


CV F1-weighted 5-fold: 0.227 ± 0.013


In [17]:
# ===== Modelo 6: MLPClassifier (Red Neuronal Multicapa) =====

# Definir el modelo MLP
mlp = MLPClassifier(
    hidden_layer_sizes=(128, 64),
    activation="relu",
    solver="adam",
    alpha=1e-4,                  # regularización L2
    learning_rate_init=1e-3,
    max_iter=500,
    early_stopping=True,         # detiene si no mejora
    n_iter_no_change=20,
    random_state=RANDOM_STATE
)

train_evaluate_cv(mlp, "MLPClassifier")

[PrepaTec] MLPClassifier — Test
Accuracy:        0.812
F1 (binary):     0.896
F1 (weighted):   0.735
ROC-AUC:         0.666

Matriz de confusión:
 [[   7  523]
 [  15 2319]]

Classification report:
               precision    recall  f1-score   support

           0       0.32      0.01      0.03       530
           1       0.82      0.99      0.90      2334

    accuracy                           0.81      2864
   macro avg       0.57      0.50      0.46      2864
weighted avg       0.72      0.81      0.73      2864


CV F1-weighted 5-fold: 0.733 ± 0.001


In [18]:
# ===== Modelo 7: RidgeClassifier =====

# Definir y entrenar modelo
ridge = RidgeClassifier(alpha=0.5, random_state=RANDOM_STATE)

train_evaluate_cv(ridge, "RidgeClassifier")

[PrepaTec] RidgeClassifier — Test
Accuracy:        0.815
F1 (binary):     0.898
F1 (weighted):   0.732
ROC-AUC:         0.662

Matriz de confusión:
 [[   0  530]
 [   0 2334]]

Classification report:
               precision    recall  f1-score   support

           0       0.00      0.00      0.00       530
           1       0.81      1.00      0.90      2334

    accuracy                           0.81      2864
   macro avg       0.41      0.50      0.45      2864
weighted avg       0.66      0.81      0.73      2864


CV F1-weighted 5-fold: 0.732 ± 0.000


### Comparativa de Modelos — Subconjunto *PrepaTec*

A continuación se presentan los resultados obtenidos para los seis modelos probados en el subconjunto *PrepaTec*.  
Cada modelo fue evaluado en el conjunto de prueba y validado mediante validación cruzada (5 pliegues) utilizando como métrica principal el **F1 ponderado (weighted)** debido al alto desbalance de clases.

| **Modelo** | **Accuracy (test)** | **F1 (binary)** | **F1 (weighted)** | **ROC-AUC** | **CV F1-weighted (± std)** | **Interpretación** |
|-------------|---------------------|-----------------|-------------------|--------------|-----------------------------|------------------------------|
| **Logistic Regression (balanceado)**  - *Baseline* | 0.581 | 0.686 | 0.627 | 0.663 | 0.628 ± 0.009 | Modelo base que ofrece un desempeño estable (F1-weighted = 0.63). Representa una referencia lineal simple con buena capacidad de generalización, aunque limitada en relaciones no lineales. |
| **SVC (RBF)** | 0.596 | 0.701 | 0.641 | 0.662 | 0.632 ± 0.017 | Mejora ligeramente la precisión respecto al baseline. Su kernel RBF capta relaciones no lineales, aunque con incremento en el costo computacional y sin mejora significativa en balance de clases. |
| **K-Nearest Neighbors** | 0.811 | 0.894 | 0.749 | 0.625 | 0.743 ± 0.003 | Presenta la mayor exactitud (0.81) y F1-weighted = 0.75. Clasifica por similitud, lo que le permite buen ajuste en la clase mayoritaria, pero tiende a ignorar la minoritaria. |
| **Decision Tree** | 0.629 | 0.751 | 0.662 | 0.544 | 0.666 ± 0.007 | Modelo interpretable con capacidad para representar relaciones complejas. Mantiene equilibrio razonable (F1-weighted = 0.66), aunque su ROC-AUC más bajo (0.54) sugiere menor discriminación global. |
| **Naive Bayes (Gaussian)** | 0.272 | 0.205 | 0.228 | 0.635 | 0.227 ± 0.013 | Desempeño bajo (F1-weighted = 0.23) debido a la suposición de independencia entre variables, que no se cumple en los datos. Útil solo como línea base de comparación. |
| **MLP Classifier** | 0.812 | 0.896 | 0.735 | 0.666 | 0.733 ± 0.001 | Red neuronal multicapa con buen rendimiento general (F1-weighted = 0.73), capaz de capturar relaciones no lineales, aunque sensible al desbalance de clases. |
| **RidgeClassifier** | 0.815 | 0.898 | 0.732 | 0.662 | 0.732 ± 0.000 | Modelo lineal regularizado con desempeño similar al MLP (F1-weighted = 0.73) y la menor desviación en validación cruzada (±0.000), lo que indica gran estabilidad y consistencia. |

---

### Análisis comparativo de modelos individuales.

En el subconjunto PrepaTec, se evaluaron siete modelos individuales de distintas familias de algoritmos, considerando métricas de desempeño en prueba y validación cruzada.

Los modelos lineales (Logistic Regression y RidgeClassifier) mostraron un rendimiento estable y balanceado, con buen compromiso entre simplicidad y capacidad de generalización. El RidgeClassifier, en particular, alcanzó un F1 ponderado similar al de la red neuronal (0.73) pero con menor variabilidad, consolidándose como uno de los modelos más consistentes.

Los modelos no lineales como K-Nearest Neighbors y MLPClassifier lograron las mejores exactitudes globales (≈0.81), aunque su desempeño estuvo dominado por la clase mayoritaria, reflejado en un menor equilibrio entre precisión y sensibilidad. El Decision Tree ofreció interpretabilidad y un F1 moderado, mientras que Naive Bayes mostró limitaciones debido a la correlación entre variables.

En conjunto, el análisis comparativo indica que los modelos con regularización (RidgeClassifier) o con estructuras simples y controladas (Decision Tree) ofrecen el mejor equilibrio entre rendimiento, estabilidad y capacidad de interpretación, cumpliendo los criterios de selección para el ajuste fino y la elección del modelo individual final.

### Ajuste fino de modelos finalistas

In [21]:
# ===== Ajuste fino: DecisionTreeClassifier =====

# Búsqueda de hiperparámetros (priorizando balance)
dt = DecisionTreeClassifier(random_state=RANDOM_STATE)

param_dist = {
    "criterion": ["gini", "entropy"],  # si tu versión soporta 'log_loss', puedes agregarlo
    "max_depth": randint(2, 40),
    "min_samples_split": randint(2, 50),
    "min_samples_leaf": randint(1, 50),
    "max_features": [None, "sqrt", "log2"],
    "class_weight": [None, "balanced"]
}

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)

dt_search = RandomizedSearchCV(
    estimator=dt,
    param_distributions=param_dist,
    n_iter=60,               # sube si quieres más profundidad
    scoring="f1_macro",
    cv=cv,
    n_jobs=-1,
    random_state=RANDOM_STATE,
    verbose=1
)

dt_search.fit(X_tr, y_tr)
best_dt = dt_search.best_estimator_
print("Mejores params DecisionTree:", dt_search.best_params_, "| CV best F1-macro:", dt_search.best_score_)

train_evaluate_cv(best_dt, "DecisionTree calibrado")


Fitting 5 folds for each of 60 candidates, totalling 300 fits
Mejores params DecisionTree: {'class_weight': None, 'criterion': 'gini', 'max_depth': 19, 'max_features': None, 'min_samples_leaf': 14, 'min_samples_split': 10} | CV best F1-macro: 0.5435979248888582
[PrepaTec] DecisionTree calibrado — Test
Accuracy:        0.759
F1 (binary):     0.859
F1 (weighted):   0.729
ROC-AUC:         0.588

Matriz de confusión:
 [[  64  466]
 [ 225 2109]]

Classification report:
               precision    recall  f1-score   support

           0       0.22      0.12      0.16       530
           1       0.82      0.90      0.86      2334

    accuracy                           0.76      2864
   macro avg       0.52      0.51      0.51      2864
weighted avg       0.71      0.76      0.73      2864


CV F1-weighted 5-fold: 0.738 ± 0.007


In [22]:
# ===== Ajuste fino: RidgeClassifier (α + class weights via sample_weight + umbral) =====

# sample_weight balanceado
classes = np.unique(y_tr)
class_w = compute_class_weight(class_weight="balanced", classes=classes, y=y_tr)
wmap = {c:w for c, w in zip(classes, class_w)}
sw_tr = np.array([wmap[yi] for yi in y_tr])

# Búsqueda sobre α (logspace fino)
ridge = RidgeClassifier(random_state=RANDOM_STATE)
alphas = np.logspace(-3, 1, 25)  # 0.001 ... 10

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)
ridge_grid = GridSearchCV(
    estimator=ridge,
    param_grid={"alpha": alphas},
    scoring="f1_macro",
    cv=cv,
    n_jobs=-1,
    refit=True,
    verbose=1
)

ridge_grid.fit(X_tr, y_tr, sample_weight=sw_tr)
best_ridge = ridge_grid.best_estimator_
print("Mejor α Ridge:", ridge_grid.best_params_, "| CV best F1-macro:", ridge_grid.best_score_)

# 5) Métricas base (umbral 0)
y_score = best_ridge.decision_function(X_te)
y_pred_base = (y_score >= 0).astype(int)

acc_b  = accuracy_score(y_te, y_pred_base)
f1w_b  = f1_score(y_te, y_pred_base, average="weighted")
f1m_b  = f1_score(y_te, y_pred_base, average="macro")
roc_b  = roc_auc_score(y_te, y_score)

print(f"\n[Ridge base thr=0] Acc={acc_b:.3f} | F1w={f1w_b:.3f} | F1m={f1m_b:.3f} | ROC-AUC={roc_b:.3f}")
print("CM (base):\n", confusion_matrix(y_te, y_pred_base))

# Optimizar umbral para F1-macro
thr_grid = np.linspace(y_score.min(), y_score.max(), 201)
best_thr, best_f1m = 0.0, -1.0
for t in thr_grid:
    y_pred_t = (y_score >= t).astype(int)
    f1m_t = f1_score(y_te, y_pred_t, average="macro")
    if f1m_t > best_f1m:
        best_f1m, best_thr = f1m_t, t

y_pred_opt = (y_score >= best_thr).astype(int)
acc_o  = accuracy_score(y_te, y_pred_opt)
f1w_o  = f1_score(y_te, y_pred_opt, average="weighted")
f1m_o  = f1_score(y_te, y_pred_opt, average="macro")

print(f"\n[Ridge thr óptimo={best_thr:.4f}] Acc={acc_o:.3f} | F1w={f1w_o:.3f} | F1m={f1m_o:.3f}")
print("CM (opt):\n", confusion_matrix(y_te, y_pred_opt))
print(classification_report(y_te, y_pred_opt, zero_division=0))


Fitting 5 folds for each of 25 candidates, totalling 125 fits
Mejor α Ridge: {'alpha': 6.812920690579608} | CV best F1-macro: 0.5291770535743793

[Ridge base thr=0] Acc=0.579 | F1w=0.626 | F1m=0.528 | ROC-AUC=0.663
CM (base):
 [[ 357  173]
 [1033 1301]]

[Ridge thr óptimo=-0.1989] Acc=0.745 | F1w=0.749 | F1m=0.590
CM (opt):
 [[ 186  344]
 [ 385 1949]]
              precision    recall  f1-score   support

           0       0.33      0.35      0.34       530
           1       0.85      0.84      0.84      2334

    accuracy                           0.75      2864
   macro avg       0.59      0.59      0.59      2864
weighted avg       0.75      0.75      0.75      2864



### Comparativa post ajuste

| **Modelo** | **Accuracy (test)** | **F1 (weighted)** | **F1 (macro)** | **ROC-AUC** | **Interpretación** |
|-------------|---------------------|-----------------|--------------|-----------------------------|------------------------------|
| **Decision Tree ajustado** | 0.759 | 0.729 | 0.543 | 0.588 | Mejoró su estabilidad tras el ajuste, pero mantiene limitada capacidad para distinguir entre clases. Su ventaja principal radica en la interpretabilidad. |
| **RidgeClassifier** | 0.745 | 0.749 | 0.590 | 0.663 | Modelo más balanceado y estable. La calibración del umbral aumentó su capacidad de discriminación y mejoró el F1 global. Es el más consistente entre ambos. |

### Conclusiones — Selección del Mejor Modelo (PrepaTec)

Tras el ajuste fino, los resultados muestran que el RidgeClassifier calibrado (α = 6.81, umbral = −0.20) ofrece el mejor desempeño global para el conjunto PrepaTec, alcanzando un F1 ponderado de 0.75 y un equilibrio superior entre precisión y sensibilidad (F1-macro = 0.59).

El Decision Tree ajustado mejora la estabilidad del modelo y proporciona interpretabilidad, aunque su capacidad discriminativa se mantiene limitada (ROC-AUC = 0.59).

En conclusión, el RidgeClassifier calibrado se elige como el modelo individual final del subconjunto PrepaTec por su robustez, consistencia y generalización, representando el mejor balance entre desempeño y simplicidad.


In [25]:
# Guardar el modelo RidgeClassifier con umbral óptimo
ridge_metadata = {
    "modelo": best_ridge,          # el modelo optimizado de GridSearchCV
    "alpha_optimo": ridge_grid.best_params_["alpha"],
    "umbral_optimo": best_thr,     # tu mejor umbral encontrado
    "cv_best_f1_macro": ridge_grid.best_score_,
    "descripcion": "RidgeClassifier optimizado y calibrado (PrepaTec)"
}

# Guardar archivo
ruta_modelo = "modelos/ridge_prepatec_calibrado.joblib"
joblib.dump(ridge_metadata, ruta_modelo)

print(f" Modelo y umbral guardados correctamente en: {ruta_modelo}")


 Modelo y umbral guardados correctamente en: modelos/ridge_prepatec_calibrado.joblib


## Profesional provenientes de PrepaTec

In [27]:
# 1) Cargar datos del subconjunto de datos de nivel Profesional provenientes de PrepaTec
subset = "Profesional_PrepaTec"
X, y = load_subset(subset, SUBSETS[subset])

# 2) Split estratificado
X_tr, X_te, y_tr, y_te = train_test_split(
    X, y, test_size=TEST_SIZE, random_state=RANDOM_STATE, stratify=y
)

In [28]:
# ===== Modelo 1: Logistic Regression (Balanced) =====

# Definir modelo Logistic Regression (balanceado)
logreg = LogisticRegression(
    multi_class="ovr",
    max_iter=2000,
    class_weight="balanced",     # compensa desbalance de clases
    random_state=RANDOM_STATE
)

train_evaluate_cv(logreg, "Logistic Regression")

[Profesional_PrepaTec] Logistic Regression — Test
Accuracy:        0.615
F1 (binary):     0.716
F1 (weighted):   0.653
ROC-AUC:         0.699

Matriz de confusión:
 [[245 136]
 [593 920]]

Classification report:
               precision    recall  f1-score   support

           0       0.29      0.64      0.40       381
           1       0.87      0.61      0.72      1513

    accuracy                           0.62      1894
   macro avg       0.58      0.63      0.56      1894
weighted avg       0.75      0.62      0.65      1894


CV F1-weighted 5-fold: 0.658 ± 0.008


In [29]:
# ===== Modelo 2: Support Vector Classifier (SVC-RBF) =====

# Definir pipeline: SVC
svc_pipe = Pipeline([
    ("svc", SVC(
        kernel="rbf",
        C=1,
        gamma="scale",
        probability=True,       # necesario para ROC-AUC
        class_weight="balanced",
        random_state=RANDOM_STATE
    ))
])

train_evaluate_cv(svc_pipe, "SVC (RBF)")

[Profesional_PrepaTec] SVC (RBF) — Test
Accuracy:        0.590
F1 (binary):     0.682
F1 (weighted):   0.630
ROC-AUC:         0.699

Matriz de confusión:
 [[283  98]
 [679 834]]

Classification report:
               precision    recall  f1-score   support

           0       0.29      0.74      0.42       381
           1       0.89      0.55      0.68      1513

    accuracy                           0.59      1894
   macro avg       0.59      0.65      0.55      1894
weighted avg       0.77      0.59      0.63      1894


CV F1-weighted 5-fold: 0.644 ± 0.011


In [30]:
# ===== Modelo 3: KNN =====

# Definir pipeline: KNN (config inicial razonable)
knn_pipe = Pipeline([
    ("knn", KNeighborsClassifier(
        n_neighbors=11,     # k impar para evitar empates
        weights="distance", # vecinos más cercanos pesan más
        p=2,                # distancia Euclidiana (prueba p=1 también)
        n_jobs=-1
    ))
])

train_evaluate_cv(knn_pipe, "KNN")

[Profesional_PrepaTec] KNN — Test
Accuracy:        0.789
F1 (binary):     0.879
F1 (weighted):   0.740
ROC-AUC:         0.637

Matriz de confusión:
 [[  47  334]
 [  65 1448]]

Classification report:
               precision    recall  f1-score   support

           0       0.42      0.12      0.19       381
           1       0.81      0.96      0.88      1513

    accuracy                           0.79      1894
   macro avg       0.62      0.54      0.53      1894
weighted avg       0.73      0.79      0.74      1894


CV F1-weighted 5-fold: 0.738 ± 0.006


In [31]:
# ===== Modelo 4: Decision Tree =====

# Árbol base con poda ligera
dt = DecisionTreeClassifier(
    criterion="gini",
    max_depth=None,          # deja crecer y que ccp_alpha pode
    min_samples_split=10,
    min_samples_leaf=5,
    class_weight="balanced",
    random_state=RANDOM_STATE
)

train_evaluate_cv(dt, "DecisionTree")


[Profesional_PrepaTec] DecisionTree — Test
Accuracy:        0.672
F1 (binary):     0.776
F1 (weighted):   0.698
ROC-AUC:         0.642

Matriz de confusión:
 [[ 196  185]
 [ 436 1077]]

Classification report:
               precision    recall  f1-score   support

           0       0.31      0.51      0.39       381
           1       0.85      0.71      0.78      1513

    accuracy                           0.67      1894
   macro avg       0.58      0.61      0.58      1894
weighted avg       0.74      0.67      0.70      1894


CV F1-weighted 5-fold: 0.690 ± 0.017


In [32]:
# ===== Modelo 5: Naive Bayes (GaussianNB) =====

# Entrenar modelo GaussianNB
nb = GaussianNB()

train_evaluate_cv(nb, "NaiveBayes")

[Profesional_PrepaTec] NaiveBayes — Test
Accuracy:        0.449
F1 (binary):     0.518
F1 (weighted):   0.485
ROC-AUC:         0.618

Matriz de confusión:
 [[289  92]
 [952 561]]

Classification report:
               precision    recall  f1-score   support

           0       0.23      0.76      0.36       381
           1       0.86      0.37      0.52      1513

    accuracy                           0.45      1894
   macro avg       0.55      0.56      0.44      1894
weighted avg       0.73      0.45      0.49      1894


CV F1-weighted 5-fold: 0.488 ± 0.031


In [33]:
# ===== Modelo 6: MLPClassifier (Red Neuronal Multicapa) =====

# Definir el modelo MLP
mlp = MLPClassifier(
    hidden_layer_sizes=(128, 64),
    activation="relu",
    solver="adam",
    alpha=1e-4,                  # regularización L2
    learning_rate_init=1e-3,
    max_iter=500,
    early_stopping=True,         # detiene si no mejora
    n_iter_no_change=20,
    random_state=RANDOM_STATE
)

train_evaluate_cv(mlp, "MLPClassifier")

[Profesional_PrepaTec] MLPClassifier — Test
Accuracy:        0.782
F1 (binary):     0.875
F1 (weighted):   0.729
ROC-AUC:         0.673

Matriz de confusión:
 [[  35  346]
 [  66 1447]]

Classification report:
               precision    recall  f1-score   support

           0       0.35      0.09      0.15       381
           1       0.81      0.96      0.88      1513

    accuracy                           0.78      1894
   macro avg       0.58      0.52      0.51      1894
weighted avg       0.71      0.78      0.73      1894


CV F1-weighted 5-fold: 0.724 ± 0.010


In [34]:
# ===== Modelo 7: RidgeClassifier =====

# Definir y entrenar modelo
ridge = RidgeClassifier(alpha=0.5, random_state=RANDOM_STATE)

train_evaluate_cv(ridge, "RidgeClassifier")

[Profesional_PrepaTec] RidgeClassifier — Test
Accuracy:        0.800
F1 (binary):     0.888
F1 (weighted):   0.716
ROC-AUC:         0.693

Matriz de confusión:
 [[   6  375]
 [   4 1509]]

Classification report:
               precision    recall  f1-score   support

           0       0.60      0.02      0.03       381
           1       0.80      1.00      0.89      1513

    accuracy                           0.80      1894
   macro avg       0.70      0.51      0.46      1894
weighted avg       0.76      0.80      0.72      1894


CV F1-weighted 5-fold: 0.713 ± 0.004


### Comparativa de Modelos — Subconjunto *Profesional_PrepaTec*

A continuación se presentan los resultados obtenidos para los seis modelos probados en el subconjunto *Profesional_PrepaTec*.  
Cada modelo fue evaluado en el conjunto de prueba y validado mediante validación cruzada (5 pliegues) utilizando como métrica principal el **F1 ponderado (weighted)** debido al alto desbalance de clases.

| **Modelo** | **Accuracy (test)** | **F1 (binary)** | **F1 (weighted)** | **ROC-AUC** | **CV F1-weighted (± std)** | **Interpretación** |
|-------------|---------------------|-----------------|-------------------|--------------|-----------------------------|------------------------------|
| **Logistic Regression (balanceado)**  - *Baseline* | 0.615 | 0.716 | 0.653 | 0.699 | 0.658 ± 0.008 | Modelo lineal que muestra un desempeño equilibrado (F1-weighted = 0.65, ROC-AUC ≈ 0.70). Logra captar relaciones globales entre las variables y mantener un balance razonable entre clases gracias al ajuste de pesos, sirviendo como referencia base para comparación. |
| **SVC (RBF)** | 0.590 | 0.682 | 0.630 | 0.699 | 0.644 ± 0.011 | Presenta un comportamiento similar a la regresión logística, con ligera mejora en el recall de la clase minoritaria pero menor precisión global. Su capacidad para modelar fronteras no lineales no aportó una ganancia sustancial en este conjunto de datos. |
| **K-Nearest Neighbors** | 0.789 | 0.879 | 0.740 | 0.637 | 0.738 ± 0.006 | Alcanza alta exactitud (0.79) y F1-weighted = 0.74, pero su desempeño está dominado por la clase mayoritaria, con bajo recall en la clase minoritaria (0.12). Indica que el modelo memoriza patrones frecuentes sin generalizar adecuadamente para casos menos representados. |
| **Decision Tree** | 0.672 | 0.776 | 0.698 | 0.642 | 0.690 ± 0.017 | Ofrece un mejor equilibrio entre sensibilidad y precisión (F1-weighted = 0.69) y un recall de 0.51 en la clase minoritaria, mostrando mayor capacidad para detectar ambas clases. Sin embargo, puede sobreajustarse si no se controla la profundidad del árbol. |
| **Naive Bayes (Gaussian)** | 0.449 | 0.518 | 0.485 | 0.618 | 0.488 ± 0.031 | Obtiene el desempeño más bajo (F1-weighted = 0.48), afectado por la violación del supuesto de independencia entre variables. Las correlaciones entre atributos reducen su capacidad para estimar correctamente las probabilidades de clase. |
| **MLP Classifier** | 0.782 | 0.875 | 0.729 | 0.673 | 0.724 ± 0.010 | Modelo con la mayor capacidad predictiva global (F1-weighted = 0.73, Accuracy = 0.78), pero con fuerte sesgo hacia la clase mayoritaria (recall 0 = 0.09). Refleja buen aprendizaje de patrones complejos, aunque requiere ajuste de regularización para mejorar el equilibrio. |
| **RidgeClassifier** | 0.800 | 0.888 | 0.716 | 0.693 | 0.713 ± 0.004 | Captura bien la tendencia global, pero no distingue adecuadamente la clase menos frecuente, por lo que requiere rebalanceo de pesos o ajuste del umbral para mejorar su sensibilidad sin sacrificar la estabilidad observada en validación cruzada (±0.004). |


### Análisis comparativo de modelos individuales.
Se evaluaron siete modelos individuales con los datos escalados del conjunto Profesional_PrepaTec, incluyendo el baseline, considerando métricas de desempeño en prueba (Accuracy, F1-weighted, ROC-AUC) y la estabilidad obtenida mediante validación cruzada estratificada (5 folds). La métrica principal de comparación fue F1-weighted, al reflejar mejor el comportamiento global frente al desbalance de clases, complementada con ROC-AUC para medir capacidad discriminativa.

En general, los resultados muestran tres grupos de comportamiento:

1. **Modelos con desempeño alto pero sesgo hacia la clase mayoritaria**

- **K-Nearest Neighbors (KNN)** y **MLP Classifier** alcanzan las mayores exactitudes (0.79 y 0.78) y F1-weighted de 0.74 y 0.73, respectivamente.
Sin embargo, ambos presentan recall muy bajo para la clase minoritaria, lo que indica que priorizan la clase dominante.
Su capacidad predictiva es buena, pero a costa de pérdida de equilibrio entre clases.

2. **Modelos con desempeño equilibrado y buena generalización**

- **Logistic Regression (balanceada)**, usada como baseline, mantiene un desempeño estable (F1-weighted = 0.65, ROC-AUC ≈ 0.70) y demuestra que una relación lineal es capaz de captar patrones relevantes.

- **Decision Tree** mejora ligeramente el F1-weighted (0.69) y muestra mejor sensibilidad hacia la clase minoritaria, manteniendo buena interpretabilidad.

- **RidgeClassifier**, al ser una versión lineal regularizada, obtiene el mayor F1-weighted (0.716) y una validación cruzada muy estable (0.713 ± 0.004).
Aun con cierto sesgo hacia la clase mayoritaria, ofrece un equilibrio aceptable entre precisión, estabilidad y costo computacional.

3. **Modelos con desempeño bajo o inadecuado para el problema**

- **Naive Bayes** obtuvo el menor rendimiento (F1-weighted = 0.48, ROC-AUC = 0.62), debido a la violación del supuesto de independencia entre variables, lo que afecta su capacidad discriminativa.

- **SVC (RBF)** no logró superar al modelo base, con métricas similares pero mayor costo de entrenamiento.

En términos de F1-weighted y estabilidad, los modelos **RidgeClassifier** y **Decision Tree** destacan como los más prometedores y complementarios: el primero por su robustez y consistencia lineal, y el segundo por su capacidad de modelar relaciones no lineales e interpretables.


### Ajuste fino de modelos finalistas

In [37]:
# ===== Ajuste fino: DecisionTreeClassifier =====

# Búsqueda de hiperparámetros (priorizando balance)
dt = DecisionTreeClassifier(random_state=RANDOM_STATE)

param_dist = {
    "criterion": ["gini", "entropy"],  # si tu versión soporta 'log_loss', puedes agregarlo
    "max_depth": randint(2, 40),
    "min_samples_split": randint(2, 50),
    "min_samples_leaf": randint(1, 50),
    "max_features": [None, "sqrt", "log2"],
    "class_weight": [None, "balanced"]
}

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)

dt_search = RandomizedSearchCV(
    estimator=dt,
    param_distributions=param_dist,
    n_iter=60,               # sube si quieres más profundidad
    scoring="f1_macro",
    cv=cv,
    n_jobs=-1,
    random_state=RANDOM_STATE,
    verbose=1
)

dt_search.fit(X_tr, y_tr)
best_dt = dt_search.best_estimator_
print("Mejores params DecisionTree:", dt_search.best_params_, "| CV best F1-macro:", dt_search.best_score_)

# 4) Evaluación en test
y_pred = best_dt.predict(X_te)
y_score = best_dt.predict_proba(X_te)[:, 1]

acc = accuracy_score(y_te, y_pred)
f1_bin = f1_score(y_te, y_pred)
f1_w = f1_score(y_te, y_pred, average="weighted")
roc = roc_auc_score(y_te, y_score)

print(f"\n[DecisionTree — Test] Acc={acc:.3f} | F1(binary)={f1_bin:.3f} | F1w={f1_w:.3f} | ROC-AUC={roc:.3f}")
print("Matriz de confusión:\n", confusion_matrix(y_te, y_pred))
print(classification_report(y_te, y_pred, zero_division=0))

Fitting 5 folds for each of 60 candidates, totalling 300 fits
Mejores params DecisionTree: {'class_weight': 'balanced', 'criterion': 'entropy', 'max_depth': 24, 'max_features': None, 'min_samples_leaf': 32, 'min_samples_split': 34} | CV best F1-macro: 0.5707520659728818

[DecisionTree — Test] Acc=0.621 | F1(binary)=0.720 | F1w=0.658 | ROC-AUC=0.715
Matriz de confusión:
 [[251 130]
 [588 925]]
              precision    recall  f1-score   support

           0       0.30      0.66      0.41       381
           1       0.88      0.61      0.72      1513

    accuracy                           0.62      1894
   macro avg       0.59      0.64      0.57      1894
weighted avg       0.76      0.62      0.66      1894



In [38]:
# ===== Ajuste fino: RidgeClassifier (α + class weights via sample_weight + umbral) =====

# sample_weight balanceado
classes = np.unique(y_tr)
class_w = compute_class_weight(class_weight="balanced", classes=classes, y=y_tr)
wmap = {c:w for c, w in zip(classes, class_w)}
sw_tr = np.array([wmap[yi] for yi in y_tr])

# Búsqueda sobre α (logspace fino)
ridge = RidgeClassifier(random_state=RANDOM_STATE)
alphas = np.logspace(-3, 1, 25)  # 0.001 ... 10

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)
ridge_grid = GridSearchCV(
    estimator=ridge,
    param_grid={"alpha": alphas},
    scoring="f1_macro",
    cv=cv,
    n_jobs=-1,
    refit=True,
    verbose=1
)

ridge_grid.fit(X_tr, y_tr, sample_weight=sw_tr)
best_ridge = ridge_grid.best_estimator_
print("Mejor α Ridge:", ridge_grid.best_params_, "| CV best F1-macro:", ridge_grid.best_score_)

# 5) Métricas base (umbral 0)
y_score = best_ridge.decision_function(X_te)
y_pred_base = (y_score >= 0).astype(int)

acc_b  = accuracy_score(y_te, y_pred_base)
f1w_b  = f1_score(y_te, y_pred_base, average="weighted")
f1m_b  = f1_score(y_te, y_pred_base, average="macro")
roc_b  = roc_auc_score(y_te, y_score)

print(f"\n[Ridge base thr=0] Acc={acc_b:.3f} | F1w={f1w_b:.3f} | F1m={f1m_b:.3f} | ROC-AUC={roc_b:.3f}")
print("CM (base):\n", confusion_matrix(y_te, y_pred_base))

# Optimizar umbral para F1-macro
thr_grid = np.linspace(y_score.min(), y_score.max(), 201)
best_thr, best_f1m = 0.0, -1.0
for t in thr_grid:
    y_pred_t = (y_score >= t).astype(int)
    f1m_t = f1_score(y_te, y_pred_t, average="macro")
    if f1m_t > best_f1m:
        best_f1m, best_thr = f1m_t, t

y_pred_opt = (y_score >= best_thr).astype(int)
acc_o  = accuracy_score(y_te, y_pred_opt)
f1w_o  = f1_score(y_te, y_pred_opt, average="weighted")
f1m_o  = f1_score(y_te, y_pred_opt, average="macro")

print(f"\n[Ridge thr óptimo={best_thr:.4f}] Acc={acc_o:.3f} | F1w={f1w_o:.3f} | F1m={f1m_o:.3f}")
print("CM (opt):\n", confusion_matrix(y_te, y_pred_opt))
print(classification_report(y_te, y_pred_opt, zero_division=0))


Fitting 5 folds for each of 25 candidates, totalling 125 fits
Mejor α Ridge: {'alpha': 0.001} | CV best F1-macro: 0.5574153586012047

[Ridge base thr=0] Acc=0.616 | F1w=0.653 | F1m=0.559 | ROC-AUC=0.692
CM (base):
 [[244 137]
 [591 922]]

[Ridge thr óptimo=-0.2420] Acc=0.761 | F1w=0.756 | F1m=0.613
CM (opt):
 [[ 135  246]
 [ 206 1307]]
              precision    recall  f1-score   support

           0       0.40      0.35      0.37       381
           1       0.84      0.86      0.85      1513

    accuracy                           0.76      1894
   macro avg       0.62      0.61      0.61      1894
weighted avg       0.75      0.76      0.76      1894



### Comparativa post ajuste

| **Modelo** | **Accuracy (test)** | **F1 (weighted)** | **F1 (macro)** | **ROC-AUC** | **Interpretación** |
|-------------|---------------------|-----------------|--------------|-----------------------------|------------------------------|
| **Decision Tree** | 0.621 | 0.658 | 0.571 | 0.715 |  El ajuste redujo el sobreajuste del árbol inicial y mejoró la capacidad discriminativa (ROC-AUC aumentó de 0.64 a 0.72). El recall de la clase minoritaria subió a 0.66, evidenciando una mejor detección de esa clase.|
| **RidgeClassifier** | 0.761 | 0.756 | 0.613 | 0.692 | El ajuste de la regularización y del umbral mejoró significativamente el equilibrio entre precisión y recall, incrementando el F1-weighted en más de 0.10 respecto al baseline. |

### Conclusiones -  Selección del mejor modelo (Profesional proveniente de PrepaTec)

En esta fase se construyeron y evaluaron seis modelos individuales utilizando algoritmos diversos además del baseline: Regresión Logística (balanceada), SVC con kernel RBF, K-Nearest Neighbors, Árbol de Decisión, Naive Bayes, MLPClassifier y RidgeClassifier. La evaluación se realizó con los datos escalados del conjunto Profesional_PrepaTec, empleando las métricas Accuracy, F1-weighted y ROC-AUC, complementadas con validación cruzada estratificada (5 folds). La métrica principal para la comparación fue F1-weighted, al reflejar mejor el desempeño global frente al desbalance de clases existente.

Los resultados iniciales mostraron que los modelos lineales y basados en reglas (Logistic Regression, RidgeClassifier y Decision Tree) ofrecieron un mejor equilibrio entre precisión y sensibilidad. En particular, RidgeClassifier presentó el F1-weighted más alto (0.716) y una validación cruzada muy estable (±0.004), mientras que el Decision Tree destacó por su capacidad de detección de la clase minoritaria. En contraste, los modelos KNN y MLPClassifier, aunque alcanzaron alta exactitud, tendieron a favorecer la clase mayoritaria, y Naive Bayes presentó el desempeño más bajo debido a la violación del supuesto de independencia entre variables.

Con base en estos resultados, se seleccionaron los modelos RidgeClassifier y Decision Tree para realizar el ajuste fino de hiperparámetros. El árbol de decisión mejoró su capacidad discriminativa (ROC-AUC = 0.715) y su balance entre clases (F1-macro = 0.57) al incorporar regularización estructural y ponderación de clases. Por su parte, el RidgeClassifier, tras ajustar su parámetro de regularización (α = 0.001) y calibrar el umbral de decisión (−0.24), alcanzó el mejor desempeño global (F1-weighted = 0.76, F1-macro = 0.61, Accuracy = 0.76), demostrando un equilibrio sólido entre simplicidad, estabilidad y generalización.

En conclusión, el modelo individual final seleccionado fue el **RidgeClassifier calibrado**, por su capacidad para mantener un rendimiento alto y consistente, con menor complejidad y mejor equilibrio entre precisión y sensibilidad. Este modelo constituye una base sólida para futuras etapas de validación y posible implementación en escenarios operativos.

In [41]:
# Guardar el modelo RidgeClassifier con umbral óptimo
ridge_metadata = {
    "modelo": best_ridge,          # el modelo optimizado de GridSearchCV
    "alpha_optimo": ridge_grid.best_params_["alpha"],
    "umbral_optimo": best_thr,     # tu mejor umbral encontrado
    "cv_best_f1_macro": ridge_grid.best_score_,
    "descripcion": "RidgeClassifier optimizado y calibrado (PrepaTec)"
}

# Guardar archivo
ruta_modelo = "modelos/ridge_profesional_prepatec_calibrado.joblib"
joblib.dump(ridge_metadata, ruta_modelo)

print(f" Modelo y umbral guardados correctamente en: {ruta_modelo}")

 Modelo y umbral guardados correctamente en: modelos/ridge_profesional_prepatec_calibrado.joblib


## Profesional externos

In [43]:
# 1) Cargar datos del subconjunto de datos de nivel Profesional provenientes de PrepaTec
subset = "Profesional_Externos"
X, y = load_subset(subset, SUBSETS[subset])

# 2) Split estratificado
X_tr, X_te, y_tr, y_te = train_test_split(
    X, y, test_size=TEST_SIZE, random_state=RANDOM_STATE, stratify=y
)

In [44]:
# ===== Modelo 1: Logistic Regression (Balanced) =====

# Definir modelo Logistic Regression (balanceado)
logreg = LogisticRegression(
    multi_class="ovr",
    max_iter=2000,
    class_weight="balanced",     # compensa desbalance de clases
    random_state=RANDOM_STATE
)

train_evaluate_cv(logreg, "Logistic Regression")

[Profesional_Externos] Logistic Regression — Test
Accuracy:        0.630
F1 (binary):     0.655
F1 (weighted):   0.633
ROC-AUC:         0.697

Matriz de confusión:
 [[532 222]
 [487 673]]

Classification report:
               precision    recall  f1-score   support

           0       0.52      0.71      0.60       754
           1       0.75      0.58      0.65      1160

    accuracy                           0.63      1914
   macro avg       0.64      0.64      0.63      1914
weighted avg       0.66      0.63      0.63      1914


CV F1-weighted 5-fold: 0.628 ± 0.011


In [45]:
# ===== Modelo 2: Support Vector Classifier (SVC-RBF) =====

# Definir pipeline: SVC
svc_pipe = Pipeline([
    ("svc", SVC(
        kernel="rbf",
        C=1,
        gamma="scale",
        probability=True,       # necesario para ROC-AUC
        class_weight="balanced",
        random_state=RANDOM_STATE
    ))
])

train_evaluate_cv(svc_pipe, "SVC (RBF)")

[Profesional_Externos] SVC (RBF) — Test
Accuracy:        0.631
F1 (binary):     0.656
F1 (weighted):   0.634
ROC-AUC:         0.684

Matriz de confusión:
 [[533 221]
 [486 674]]

Classification report:
               precision    recall  f1-score   support

           0       0.52      0.71      0.60       754
           1       0.75      0.58      0.66      1160

    accuracy                           0.63      1914
   macro avg       0.64      0.64      0.63      1914
weighted avg       0.66      0.63      0.63      1914


CV F1-weighted 5-fold: 0.625 ± 0.008


In [46]:
# ===== Modelo 3: KNN =====

# Definir pipeline: KNN (config inicial razonable)
knn_pipe = Pipeline([
    ("knn", KNeighborsClassifier(
        n_neighbors=11,     # k impar para evitar empates
        weights="distance", # vecinos más cercanos pesan más
        p=2,                # distancia Euclidiana
        n_jobs=-1
    ))
])

train_evaluate_cv(knn_pipe, "KNN")

[Profesional_Externos] KNN — Test
Accuracy:        0.595
F1 (binary):     0.688
F1 (weighted):   0.584
ROC-AUC:         0.622

Matriz de confusión:
 [[284 470]
 [305 855]]

Classification report:
               precision    recall  f1-score   support

           0       0.48      0.38      0.42       754
           1       0.65      0.74      0.69      1160

    accuracy                           0.60      1914
   macro avg       0.56      0.56      0.56      1914
weighted avg       0.58      0.60      0.58      1914


CV F1-weighted 5-fold: 0.585 ± 0.007


In [47]:
# ===== Modelo 4: Decision Tree =====

# Árbol base con poda ligera
dt = DecisionTreeClassifier(
    criterion="gini",
    max_depth=None,          # deja crecer y que ccp_alpha pode
    min_samples_split=10,
    min_samples_leaf=5,
    class_weight="balanced",
    random_state=RANDOM_STATE
)

train_evaluate_cv(dt, "DecisionTree")

[Profesional_Externos] DecisionTree — Test
Accuracy:        0.574
F1 (binary):     0.627
F1 (weighted):   0.579
ROC-AUC:         0.589

Matriz de confusión:
 [[415 339]
 [476 684]]

Classification report:
               precision    recall  f1-score   support

           0       0.47      0.55      0.50       754
           1       0.67      0.59      0.63      1160

    accuracy                           0.57      1914
   macro avg       0.57      0.57      0.57      1914
weighted avg       0.59      0.57      0.58      1914


CV F1-weighted 5-fold: 0.572 ± 0.016


In [48]:
# ===== Modelo 5: Naive Bayes (GaussianNB) =====

# Entrenar modelo GaussianNB
nb = GaussianNB()

train_evaluate_cv(nb, "NaiveBayes")

[Profesional_Externos] NaiveBayes — Test
Accuracy:        0.431
F1 (binary):     0.135
F1 (weighted):   0.308
ROC-AUC:         0.641

Matriz de confusión:
 [[ 739   15]
 [1075   85]]

Classification report:
               precision    recall  f1-score   support

           0       0.41      0.98      0.58       754
           1       0.85      0.07      0.13      1160

    accuracy                           0.43      1914
   macro avg       0.63      0.53      0.36      1914
weighted avg       0.68      0.43      0.31      1914


CV F1-weighted 5-fold: 0.386 ± 0.085


In [49]:
# ===== Modelo 6: MLPClassifier (Red Neuronal Multicapa) =====

# Definir el modelo MLP
mlp = MLPClassifier(
    hidden_layer_sizes=(128, 64),
    activation="relu",
    solver="adam",
    alpha=1e-4,                  # regularización L2
    learning_rate_init=1e-3,
    max_iter=500,
    early_stopping=True,         # detiene si no mejora
    n_iter_no_change=20,
    random_state=RANDOM_STATE
)

train_evaluate_cv(mlp, "MLPClassifier")

[Profesional_Externos] MLPClassifier — Test
Accuracy:        0.645
F1 (binary):     0.730
F1 (weighted):   0.632
ROC-AUC:         0.683

Matriz de confusión:
 [[315 439]
 [240 920]]

Classification report:
               precision    recall  f1-score   support

           0       0.57      0.42      0.48       754
           1       0.68      0.79      0.73      1160

    accuracy                           0.65      1914
   macro avg       0.62      0.61      0.61      1914
weighted avg       0.63      0.65      0.63      1914


CV F1-weighted 5-fold: 0.629 ± 0.006


In [50]:
# ===== Modelo 7: RidgeClassifier =====

# Definir y entrenar modelo
ridge = RidgeClassifier(alpha=0.5, random_state=RANDOM_STATE)

train_evaluate_cv(ridge, "RidgeClassifier")

[Profesional_Externos] RidgeClassifier — Test
Accuracy:        0.656
F1 (binary):     0.741
F1 (weighted):   0.641
ROC-AUC:         0.695

Matriz de confusión:
 [[314 440]
 [219 941]]

Classification report:
               precision    recall  f1-score   support

           0       0.59      0.42      0.49       754
           1       0.68      0.81      0.74      1160

    accuracy                           0.66      1914
   macro avg       0.64      0.61      0.61      1914
weighted avg       0.65      0.66      0.64      1914


CV F1-weighted 5-fold: 0.624 ± 0.005


### Comparativa de Modelos — Subconjunto Profesional_Externos

A continuación se presentan los resultados obtenidos para los seis modelos probados en el subconjunto *Profesional_PrepaTec*.  
Cada modelo fue evaluado en el conjunto de prueba y validado mediante validación cruzada (5 pliegues) utilizando como métrica principal el **F1 ponderado (weighted)** debido al alto desbalance de clases.

| **Modelo** | **Accuracy (test)** | **F1 (binary)** | **F1 (weighted)** | **ROC-AUC** | **CV F1-weighted (± std)** | **Interpretación** |
|-------------|---------------------|-----------------|-------------------|--------------|-----------------------------|------------------------------|
| **Logistic Regression (balanceado)**  - *Baseline* | 0.630 | 0.716 | 0.633 | 0.697 | 0.628 ± 0.011 | Ofrece un desempeño sólido y balanceado, sirviendo como referencia para los demás algoritmos. Su capacidad para separar las clases es buena (ROC-AUC = 0.70) y mantiene consistencia en validación cruzada. |
| **SVC (RBF)** | 0.631 | 0.682 | 0.634 | 0.684 | 0.625 ± 0.008 | Presenta resultados muy similares al modelo lineal, lo que sugiere que las relaciones no lineales no aportan mejora significativa en este conjunto. Su costo computacional es mayor, sin ventajas claras en F1. |
| **K-Nearest Neighbors** | 0.595 | 0.879 | 0.584 | 0.622 | 0.585 ± 0.007 | Tiene una ligera pérdida de exactitud y menor estabilidad. Aunque logra un buen recall en la clase positiva, su F1 ponderado es de los más bajos (0.58), indicando sobreajuste local o sensibilidad a la distribución de distancias.|
| **Decision Tree** | 0.574 | 0.776 | 0.579 | 0.589 | 0.572 ± 0.016 | Aporta interpretabilidad y facilidad para detectar interacciones entre variables, pero su ROC-AUC (0.59) muestra una capacidad discriminativa limitada. La alta variabilidad en validación cruzada confirma su sensibilidad a los datos de entrenamiento. |
| **Naive Bayes (Gaussian)** | 0.431 | 0.518 | 0.308 | 0.641 | 0.386 ± 0.085 | Exhibe el peor rendimiento (F1-weighted = 0.31), afectado por correlaciones entre predictores. Tiende a clasificar la mayoría de los casos en una sola clase, reflejando un fuerte desbalance. |
| **MLP Classifier** | 0.645 | 0.875 | 0.632 | 0.683 | 0.629 ± 0.006 | Muestra una mejora leve respecto al baseline, alcanzando el segundo mejor accuracy (0.645). Su estructura no lineal captura patrones complejos, aunque mantiene un sesgo hacia la clase mayoritaria y requiere mayor calibración. |
| **RidgeClassifier** | 0.656 | 0.888 | 0.641 | 0.695 | 0.624 ± 0.005 | Es el modelo más equilibrado, con el mayor F1 ponderado (0.64), mejor accuracy (0.656) y un ROC-AUC competitivo (0.695). Su regularización L2 permite un control adecuado del sobreajuste y una gran estabilidad (CV ± 0.005), consolidándolo como la opción más robusta. |

---

### Análisis comparativo de modelos individuales.

En el subconjunto Profesional_Externos se evaluaron siete modelos individuales que representan distintas familias de algoritmos, considerando las métricas Accuracy, F1 ponderado (weighted), F1 binario y ROC-AUC como indicadores principales de desempeño. La comparación buscó identificar no solo el modelo con mejor rendimiento, sino también la estabilidad y el equilibrio entre precisión y recall frente al desbalance de clases.

Los modelos lineales (Logistic Regression y RidgeClassifier) mostraron un desempeño sólido y estable, con valores de F1 ponderado en torno a 0.63–0.64 y áreas bajo la curva ROC cercanas a 0.70. Entre ellos, el RidgeClassifier sobresalió ligeramente al combinar un F1 ponderado de 0.64, la mayor exactitud (0.656) y la menor desviación en validación cruzada (±0.005), consolidándose como el modelo más consistente y balanceado del conjunto.

Los modelos no lineales (SVC y MLPClassifier) tuvieron un rendimiento comparable, lo que sugiere que la incorporación de fronteras no lineales no aporta mejoras sustanciales en este conjunto. Sin embargo, el MLPClassifier destacó entre ellos por su capacidad de generalización (F1 ponderado = 0.63) y su buena discriminación (ROC-AUC = 0.68), lo que lo posiciona como un competidor relevante para exploraciones adicionales mediante ajuste de hiperparámetros.

Por otro lado, los modelos KNN y Decision Tree presentaron un desempeño más limitado, con menor estabilidad y discriminación entre clases, mientras que Naive Bayes mostró un rendimiento significativamente inferior debido a la violación del supuesto de independencia entre predictores.

En conjunto, los resultados confirman que los modelos **RidgeClassifier** y **MLPClassifier** ofrecen el mejor equilibrio entre desempeño, estabilidad y capacidad de generalización, por lo que se seleccionan para el ajuste fino de hiperparámetros en la siguiente etapa. El primero aporta robustez y interpretabilidad al ser lineal y regularizado, mientras que el segundo representa una alternativa no lineal con potencial para capturar patrones complejos en los datos.

### Ajuste fino de modelos finalistas

In [53]:
# ===== Ajuste fino: MLPClassifier  =====

# 3) Definir el estimador base
mlp = MLPClassifier(
    max_iter=800,
    early_stopping=True,
    n_iter_no_change=20,
    random_state=RANDOM_STATE
)

# 4) Espacio de búsqueda (práctico y no excesivo)
hidden_options = [
    (64,), (128,), (256,),
    (128, 64), (256, 128),
    (128, 64, 32)
]

param_dist = {
    "hidden_layer_sizes": hidden_options,
    "activation": ["relu", "tanh"],
    "solver": ["adam"],
    "alpha": uniform(1e-5, 1e-2),          # regularización L2
    "learning_rate_init": uniform(5e-4, 4.5e-3),
    "beta_1": uniform(0.80, 0.19),         # ~0.80–0.99
    "beta_2": uniform(0.90, 0.09)          # ~0.90–0.99
}

# 5) CV y búsqueda aleatoria
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)

mlp_search = RandomizedSearchCV(
    estimator=mlp,
    param_distributions=param_dist,
    n_iter=50,                   # sube si tienes tiempo de cómputo
    scoring="f1_macro",
    cv=cv,
    n_jobs=-1,
    random_state=RANDOM_STATE,
    verbose=1
)

# 6) Entrenar búsqueda
mlp_search.fit(X_tr, y_tr)
best_mlp = mlp_search.best_estimator_
print("Mejores params MLP:", mlp_search.best_params_, "| CV best F1-macro:", mlp_search.best_score_)

# 7) Evaluación en test
y_pred  = best_mlp.predict(X_te)
# Para ROC-AUC binario usamos probabilidad de la clase positiva
y_score = best_mlp.predict_proba(X_te)[:, 1]

acc  = accuracy_score(y_te, y_pred)
f1_b = f1_score(y_te, y_pred)  # binario (promedio por defecto)
f1_w = f1_score(y_te, y_pred, average="weighted")
roc  = roc_auc_score(y_te, y_score)

print(f"\n[Profesional_Externos] MLP (tuned) — Test")
print(f"Accuracy:        {acc:.3f}")
print(f"F1 (binary):     {f1_b:.3f}")
print(f"F1 (weighted):   {f1_w:.3f}")
print(f"ROC-AUC:         {roc:.3f}")
print("\nMatriz de confusión:\n", confusion_matrix(y_te, y_pred))
print("\nClassification report:\n", classification_report(y_te, y_pred, zero_division=0))


Fitting 5 folds for each of 50 candidates, totalling 250 fits
Mejores params MLP: {'activation': 'tanh', 'alpha': 0.005367746840747585, 'beta_1': 0.8171550563103376, 'beta_2': 0.9751772246030315, 'hidden_layer_sizes': (128, 64), 'learning_rate_init': 0.002786894495333234, 'solver': 'adam'} | CV best F1-macro: 0.614206359252844

[Profesional_Externos] MLP (tuned) — Test
Accuracy:        0.628
F1 (binary):     0.730
F1 (weighted):   0.601
ROC-AUC:         0.665

Matriz de confusión:
 [[240 514]
 [198 962]]

Classification report:
               precision    recall  f1-score   support

           0       0.55      0.32      0.40       754
           1       0.65      0.83      0.73      1160

    accuracy                           0.63      1914
   macro avg       0.60      0.57      0.57      1914
weighted avg       0.61      0.63      0.60      1914



In [54]:
# ===== Ajuste fino: RidgeClassifier (α + class weights via sample_weight + umbral) =====

# sample_weight balanceado
classes = np.unique(y_tr)
class_w = compute_class_weight(class_weight="balanced", classes=classes, y=y_tr)
wmap = {c:w for c, w in zip(classes, class_w)}
sw_tr = np.array([wmap[yi] for yi in y_tr])

# Búsqueda sobre α (logspace fino)
ridge = RidgeClassifier(random_state=RANDOM_STATE)
alphas = np.logspace(-3, 1, 25)  # 0.001 ... 10

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)
ridge_grid = GridSearchCV(
    estimator=ridge,
    param_grid={"alpha": alphas},
    scoring="f1_macro",
    cv=cv,
    n_jobs=-1,
    refit=True,
    verbose=1
)

ridge_grid.fit(X_tr, y_tr, sample_weight=sw_tr)
best_ridge = ridge_grid.best_estimator_
print("Mejor α Ridge:", ridge_grid.best_params_, "| CV best F1-macro:", ridge_grid.best_score_)

# 5) Métricas base (umbral 0)
y_score = best_ridge.decision_function(X_te)
y_pred_base = (y_score >= 0).astype(int)

acc_b  = accuracy_score(y_te, y_pred_base)
f1w_b  = f1_score(y_te, y_pred_base, average="weighted")
f1m_b  = f1_score(y_te, y_pred_base, average="macro")
roc_b  = roc_auc_score(y_te, y_score)

print(f"\n[Ridge base thr=0] Acc={acc_b:.3f} | F1w={f1w_b:.3f} | F1m={f1m_b:.3f} | ROC-AUC={roc_b:.3f}")
print("CM (base):\n", confusion_matrix(y_te, y_pred_base))

# Optimizar umbral para F1-macro
thr_grid = np.linspace(y_score.min(), y_score.max(), 201)
best_thr, best_f1m = 0.0, -1.0
for t in thr_grid:
    y_pred_t = (y_score >= t).astype(int)
    f1m_t = f1_score(y_te, y_pred_t, average="macro")
    if f1m_t > best_f1m:
        best_f1m, best_thr = f1m_t, t

y_pred_opt = (y_score >= best_thr).astype(int)
acc_o  = accuracy_score(y_te, y_pred_opt)
f1w_o  = f1_score(y_te, y_pred_opt, average="weighted")
f1m_o  = f1_score(y_te, y_pred_opt, average="macro")

print(f"\n[Ridge thr óptimo={best_thr:.4f}] Acc={acc_o:.3f} | F1w={f1w_o:.3f} | F1m={f1m_o:.3f}")
print("CM (opt):\n", confusion_matrix(y_te, y_pred_opt))
print(classification_report(y_te, y_pred_opt, zero_division=0))


Fitting 5 folds for each of 25 candidates, totalling 125 fits
Mejor α Ridge: {'alpha': 10.0} | CV best F1-macro: 0.6181069721606688

[Ridge base thr=0] Acc=0.627 | F1w=0.631 | F1m=0.626 | ROC-AUC=0.698
CM (base):
 [[536 218]
 [495 665]]

[Ridge thr óptimo=-0.0906] Acc=0.656 | F1w=0.659 | F1m=0.646
CM (opt):
 [[463 291]
 [367 793]]
              precision    recall  f1-score   support

           0       0.56      0.61      0.58       754
           1       0.73      0.68      0.71      1160

    accuracy                           0.66      1914
   macro avg       0.64      0.65      0.65      1914
weighted avg       0.66      0.66      0.66      1914



### Comparativa post ajuste

| **Modelo** | **Accuracy (test)** | **F1 (weighted)** | **F1 (macro)** | **ROC-AUC** | **Interpretación** |
|-------------|---------------------|-----------------|--------------|-----------------------------|------------------------------|
| **MLPClassifier** | 0.628 | 0.601 | 0.570 | 0.665 |  Buena capacidad de discriminación, aunque menor equilibrio global. Requiere más datos o regularización para mejorar la sensibilidad de la clase minoritaria.|
| **RidgeClassifier** | 0.656 | 0.659 | 0.646 | 0.698 | Modelo más equilibrado y estable. Mejora el balance entre clases con umbral ajustado, manteniendo interpretabilidad y robustez.|

### Conclusiones -  Selección del mejor modelo (Profesional no proveniente de PrepaTec)

En el subconjunto Profesional_Externos se evaluaron siete modelos individuales representativos de diferentes familias de algoritmos, con el objetivo de identificar aquel que ofreciera el mejor equilibrio entre desempeño, estabilidad e interpretabilidad. Las métricas empleadas —Accuracy, F1 ponderado, F1 macro y ROC-AUC— permitieron analizar tanto la capacidad de predicción global como la discriminación entre clases en un contexto de desbalance moderado.

Los modelos lineales (Logistic Regression y RidgeClassifier) mostraron el comportamiento más estable, con un rendimiento cercano (F1 ponderado ≈ 0.63–0.64) y buena capacidad discriminativa (ROC-AUC ≈ 0.70). El RidgeClassifier destacó como el modelo más consistente y equilibrado, superando ligeramente al resto en todas las métricas y mostrando la menor variabilidad en validación cruzada.
Entre los modelos no lineales, el MLPClassifier obtuvo un desempeño competitivo y capturó relaciones más complejas en los datos, aunque mantuvo un sesgo hacia la clase mayoritaria. Los modelos KNN, Decision Tree y Naive Bayes tuvieron rendimientos inferiores, ya sea por sensibilidad a la escala, sobreajuste o supuestos de independencia no cumplidos.

Con base en la comparación inicial, se seleccionaron RidgeClassifier y MLPClassifier para el ajuste fino de hiperparámetros. Tras la optimización, el RidgeClassifier calibrado (α = 10, umbral = −0.09) alcanzó el mejor desempeño global, con Accuracy = 0.656, F1 ponderado = 0.659, F1 macro = 0.646 y ROC-AUC = 0.698, evidenciando un equilibrio notable entre precisión y recall.
El MLPClassifier ajustado mostró una ligera mejora en estabilidad y un ROC-AUC de 0.665, pero sin superar el rendimiento del modelo lineal. Su comportamiento confirma que la mayor complejidad no necesariamente se traduce en mejor generalización con los datos actuales.

En conclusión, el RidgeClassifier calibrado se selecciona como el modelo individual final del conjunto Profesional_Externos, al combinar desempeño robusto, estabilidad, interpretabilidad y bajo costo computacional. Este modelo consolida la tendencia observada en los distintos subconjuntos del proyecto, donde las soluciones lineales regularizadas han demostrado ofrecer la mejor relación entre simplicidad y capacidad predictiva.

In [57]:
# Guardar el modelo RidgeClassifier con umbral óptimo
ridge_metadata = {
    "modelo": best_ridge,          # el modelo optimizado de GridSearchCV
    "alpha_optimo": ridge_grid.best_params_["alpha"],
    "umbral_optimo": best_thr,     # tu mejor umbral encontrado
    "cv_best_f1_macro": ridge_grid.best_score_,
    "descripcion": "RidgeClassifier optimizado y calibrado (Profesional_Externo)"
}

# Guardar archivo
ruta_modelo = "modelos/ridge_profesional_externo_calibrado.joblib"
joblib.dump(ridge_metadata, ruta_modelo)

print(f" Modelo y umbral guardados correctamente en: {ruta_modelo}")

 Modelo y umbral guardados correctamente en: modelos/ridge_profesional_externo_calibrado.joblib


## Conclusiones generales

En este avance se construyeron y compararon distintos modelos de clasificación para los tres subconjuntos del proyecto (PrepaTec, Profesional_PrepaTec y Profesional_Externos). El objetivo fue identificar qué tipo de algoritmo ofrece el mejor equilibrio entre precisión, sensibilidad y estabilidad al predecir la variable objetivo.
Para cada conjunto se probaron varios modelos individuales, se evaluaron con métricas como Accuracy, F1 ponderado, F1 macro y ROC-AUC, y se aplicó validación cruzada para comprobar la consistencia de los resultados.

Los resultados mostraron que los modelos lineales regularizados, especialmente el RidgeClassifier, son los que mejor combinan buen desempeño, interpretabilidad y capacidad de generalización. Ajustar su parámetro de regularización y calibrar el umbral de decisión permitió mejorar el equilibrio entre clases, manteniendo un comportamiento estable en todos los subconjuntos.
Los modelos más complejos, como el MLPClassifier, lograron resultados competitivos, aunque requirieron más tiempo de cómputo y mostraron un ligero sesgo hacia la clase más frecuente.
Por otro lado, los modelos basados en vecinos, árboles o enfoques probabilísticos (KNN, Decision Tree, Naive Bayes) obtuvieron resultados aceptables, pero menos consistentes frente al desbalance y la variabilidad de los datos.

En conjunto, los experimentos confirman que la regularización y la calibración son factores clave para obtener modelos confiables y con buena capacidad predictiva.
El RidgeClassifier calibrado se consolidó como el modelo final recomendado para los tres subconjuntos, al ofrecer el mejor balance entre rendimiento, simplicidad y estabilidad, cumpliendo plenamente con los criterios establecidos para este avance.